pytential/qbx/__init__.py

```diff
@@ -225,8 +225,9 @@ def fmm_level_to_order(kernel, kernel_args, tree, level):  # noqa pylint:disable
         self.geometry_data_inspector = geometry_data_inspector
 
         if cost_model is None:
-            from pytential.qbx.cost import CostModel
-            cost_model = CostModel()
+            from pytential.qbx.cost import QBXCostModel
+            queue = cl.CommandQueue(self.cl_context)
```

Comment:
I feel that having the cost model hold onto a command queue invites concurrency bugs. Is this avoidable? I know this would be a change that would ripple down into boxtree.

Reply:
> When is this queue waited on? What's the intended synchronization behavior?

I guess never :) I should have considered synchronization more. Right now I just use a single in-order queue everywhere so this problem never manifests.

> I feel that having the cost model hold onto a command queue invites concurrency bugs. Is this avoidable? I know this would be a change that would ripple down into boxtree.

I can think about some options:
1. Each `process_*` API in the cost model is async, and it will additionally accept a `queue` argument. The downside is the signature will be different between the Python version and the CL version, making code reuse harder (e.g. needs two versions of `qbx_cost_per_box/stage`).
2. Each `process_*` API in the cost model is async. We keep a context instead of queue in the CL cost model. Each `process_*` will construct a new queue from the context, and return the result with a queue attached so user can wait on.
3. Each `process_*` API in the cost model is synchronous. We keep a context instead of queue in the CL cost model. Each `process_*` will construct a new queue from the context and wait for the queue before returning the result.

Any thoughts? Other ideas?